### ISSUE TITLE
Gardenlet times out talking to the Seed API server

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#2747]
**How to categorize this issue?**

/area quality performance robustness scalability
/kind bug
/priority normal

**What happened**:
Over the past few weeks we have observed, that the gardenlet is often timing out talking to the Seed API server during Shoot reconciliation but also while renewing the leader lease, ending in quite frequent logs including `context deadline exceeded`.
We have seen this on two of our shooted Seeds (Azure in the US, Control plane running in the EU), which have a quite dynamic load (many Shoots creating/deleting in parallel).

The problematic steps in the Shoot reconciliation/deletion that fail are mostly those, that use the `defaultTimeout` of 30s and touch a lot of resources (for example `DeleteSeedMonitoring`).

**What you expected to happen**:
The gardenlet should be able to cope with many Shoots being reconciled in parallel.

**How to reproduce it (as minimally and precisely as possible)**:
What actually limits the effect of this issue is to set the `shoot.concurrentSyncs` value of the gardenlet configuration to a low value (e.g. 20). With this, the `context deadline exceeded` errors cease to pop up in the logs.

**Anything else we need to know?**:

**Environment**:

- Gardener version: >= v1.7.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: azure (?)
- Others:

/cc @timuthy @dguendisch @zanetworker @rfranzke

**Comment by rfranzke** [comment]
Can you please share more information which parts you already investigated (networking, etc.)?

**Comment by timebertt** [comment]
While opening this issue, I stumbled upon the wrong API groups in `DeleteSeedMonitoring`, e.g. `RoleBinding` is not in `core/v1`:
https://github.com/gardener/gardener/blob/1208bc82aa0c6910b858a00993ac0b2f1588c87c/pkg/operation/botanist/monitoring.go#L423

There are 5 of these in that function (see https://github.com/gardener/gardener/compare/master...timebertt:fix/api-groups) and in each iteration the dynamic rest mapper rediscover all API resources, because it gets a `NoKindMatchError`. As the dynamic rest mapper locks on rediscovery, one call will block all other Shoot reconciliations/deletions trying to rediscover some resource in the seed.
In some cases (e.g. the gardenlet running far away from its control plane like in our case), a full discovery may take about 2 seconds, so `5*2s + lock contention time` might easily exceed the default timeout of 30s.

/status under-investigation

Stuff we should check/do next:
- [x] Is this really the root cause? Which other flow steps are affected and why?
- [x] Check why the leader election might also fail? (really related to this issue or not?)
- [x] Switch from the dynamic REST mapper to a memory cached discovery client which will only be reset when instructed specifically (ref #2415, will also fix #2675)

**Comment by rfranzke** [comment]
Very interesting, thanks.

> As the dynamic rest mapper locks on rediscovery, one call will block all other Shoot reconciliations/deletions trying to rediscover some resource in the seed.

Why is that? During re-discovery, why can't we apply further resources that are already known?

To understand it correctly, this is a consequence of the client refactoring, right (earlier we were using dedicated clients to the seed per shoot reconciliation while we only have one client for all of them now)?

Anyways, very good finding, it sounds promising. Thank you!

**Comment by timebertt** [comment]
> > As the dynamic rest mapper locks on rediscovery, one call will block all other Shoot reconciliations/deletions trying to rediscover some resource in the seed.
> 
> Why is that? During re-discovery, why can't we apply further resources that are already known?

Maybe I wasn't clear enough, let me try again :). What I meant, was that one rediscovery on a single operation (e.g. deleting an unknown/non-existing Kind in the Shoot reconciliation/deletion) will block a parallel rediscovery caused by an unknown/non-existing Kind.
But I checked again just now and saw, that the c-r client will also lock when creating a new rest client (for a new/unknown resource, mainly while using the REST mapper), which will therefore also block calls to resources that are already known for the time that the REST mapper does a full rediscovery. So this is even worse than I thought originally 😞 


> To understand it correctly, this is a consequence of the client refactoring, right (earlier we were using dedicated clients to the seed per shoot reconciliation while we only have one client for all of them now)?

Well, I would say, it's a mix of sharing the same Seed client between all parallel operations and switching to the dynamic REST mapper. The problem/bug, that we try to delete non-existing resources (wrong API groups) was there before, but it was unnoticed, because we just ignore the `NoKindMatchError` and assume that it is gone. The refactoring/switch just reveals such problems/bugs/errors.

I think the main problem lays with the dynamic rest mapper and the high amount of parallel operations in the gardenlet and especially the way we deal with unknown resources. We have some places, where we assume, that an unknown resource means, that some component is not yet deployed/enabled (e.g. managed istio stuff, VPA, HVPA and so on) and deleting it can be done without side-effects. With the switch to the dynamic rest mapper, this assumption is not correct anymore (as described above) because of lock contention and can result in a really bad situation where we basically block all other requests.
I like this assumption, as it is making things less complicated and easier to understand, so I would like to keep it. But that means, that the dynamic REST mapper is just not suitable for our usecase and we definitely have to switch to a different implementation (like the memcached rest mapper), where we can control, when exactly a re-discovery happens, so it isn't done implicitly each time we are trying to delete a non-existing kind.

Long story short: I think there is nothing wrong with sharing the client/rest mapper between operation wrt this issue, but rather with the chosen rest mapper itself...
Sorry for the long reply 😄

**Comment by timebertt** [comment]
After watching the fix from #2752 over the course of the last 24 hours on our affected Seeds, the problem seems to be fixed indeed.
I haven't seen any other flow steps that were affected. Also the leader election didn't fail that frequently anymore.
So the first two action items are resolved, I would say.

**Comment by timebertt** [comment]
Regarding replacing the dynamic REST mapper, I will have to reiterate over the options we have here.
I would actually like to keep it and rather fix some scalability problems we have seen with it in upstream. But that will probably take some time and is therefore maybe no option for us short-term.

**Comment by timebertt** [comment]
> Regarding replacing the dynamic REST mapper, I will have to reiterate over the options we have here.
> I would actually like to keep it and rather fix some scalability problems we have seen with it in upstream. But that will probably take some time and is therefore maybe no option for us short-term.

I have created https://github.com/kubernetes-sigs/controller-runtime/pull/1151, which should definitely ease the problem described here.
Once it is merged and released, we can pass a custom rate limit to the DynamicRESTMapper (e.g. once every 5s), which will help us avoiding unnecessary discovery calls, while not surfacing `NoMatchErrors` when deleting non-existing CRs.
